jubatus/util/math/random/sfmt.h

```diff
@@ -0,0 +1,250 @@
+    }
+
+    //period_certification
+    int inner = 0;
+    for (int i = 0; i < 4; i++)
+      inner ^= psfmt32[i] & parity[i];
+    for (int i = 16; i > 0; i >>= 1)
+      inner ^= inner >> i;
+    inner &= 1;
+    if (inner == 1)
+      return;
+    for (int i = 0; i < 4; i++) {
+      int work = 1;
+      for (int j = 0; j < 32; j++) {
+        if ((work & parity[i]) != 0) {
+          psfmt32[i] ^= work;
+          return;
+        }
+        work = work << 1;
+      }
+    }
+  }
+  ~sfmt() {}
+
+  uint32_t next() {
+    uint32_t *psfmt32 = &(s[0].u[0]);
+    if (idx >= N32) {
+      generate_all();
+      idx = 0;
+    }
+    return psfmt32[idx++];
+  }
+
+  void fill_int_unsafe(uint32_t *out, size_t n) {
+    uint32_t *psfmt32 = &(s[0].u[0]);
+    if (idx + n <= N32) {
+      std::memcpy(out, psfmt32 + idx, n * sizeof(uint32_t));
+      idx += n;
+    } else {
+      while (n > 0) {
+        if (idx >= N32) {
+          generate_all();
+          idx = 0;
+        }
+        size_t m = (n + idx <= N32 ? n : N32 - idx);
+        std::memcpy(out, psfmt32 + idx, m * sizeof(uint32_t));
+        idx += m;
+        out += m;
+        n -= m;
+      }
+    }
+  }
+private:
+  inline static void mm_recursion(__m128i * r, __m128i a, __m128i b,
+                                  __m128i c, __m128i d) {
+#ifdef __SSE2__
+    static const uint32_t sse2_param_mask[4] __attribute__((aligned(16)))
+      = {MSK1, MSK2, MSK3, MSK4};
+    __m128i v, x, y, z;
+    y = _mm_srli_epi32(b, SR1);
+    z = _mm_srli_si128(c, SR2);
+    v = _mm_slli_epi32(d, SL1);
+    z = _mm_xor_si128(z, a);
+    z = _mm_xor_si128(z, v);
+    x = _mm_slli_si128(a, SL2);
+    y = _mm_and_si128(y, *(const __m128i*)sse2_param_mask);
+    z = _mm_xor_si128(z, x);
+    z = _mm_xor_si128(z, y);
+    *r = z;
+#else
+    __m128i x, y;
+    {
+      uint64_t th, tl, oh, ol;
+      th = ((uint64_t)a[3] << 32) | ((uint64_t)a[2]);
+      tl = ((uint64_t)a[1] << 32) | ((uint64_t)a[0]);
+      oh = th << (SL2 * 8);
+      ol = tl << (SL2 * 8);
+      oh |= tl >> (64 - SL2 * 8);
+      x[1] = (uint32_t)(ol >> 32);
+      x[0] = (uint32_t)ol;
+      x[3] = (uint32_t)(oh >> 32);
+      x[2] = (uint32_t)oh;
+    }
+    {
+      uint64_t th, tl, oh, ol;
+      th = ((uint64_t)c[3] << 32) | ((uint64_t)c[2]);
+      tl = ((uint64_t)c[1] << 32) | ((uint64_t)c[0]);
+      oh = th >> (SR2 * 8);
+      ol = tl >> (SR2 * 8);
+      ol |= th << (64 - SR2 * 8);
+      y[1] = (uint32_t)(ol >> 32);
+      y[0] = (uint32_t)ol;
+      y[3] = (uint32_t)(oh >> 32);
+      y[2] = (uint32_t)oh;
+    }
+    (*r)[0] = a[0] ^ x[0] ^ ((b[0] >> SR1) & MSK1) ^ y[0] ^ (d[0] << SL1);
+    (*r)[1] = a[1] ^ x[1] ^ ((b[1] >> SR1) & MSK2) ^ y[1] ^ (d[1] << SL1);
+    (*r)[2] = a[2] ^ x[2] ^ ((b[2] >> SR1) & MSK3) ^ y[2] ^ (d[2] << SL1);
+    (*r)[3] = a[3] ^ x[3] ^ ((b[3] >> SR1) & MSK4) ^ y[3] ^ (d[3] << SL1);
+#endif
+  }
+
+  void generate_all() {
+    uint32_t i;
+    __m128i r1, r2;
```

Comment:
doesn't need #ifdef because __m128i is defined manually in non-sse2 environment.
But I will rewrite this method. see above comments.